### PR TITLE
`Development`: Add check for REST endpoints to be in kebab case

### PR DESCRIPTION
**Pull request by SimonEntholzer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).

### Motivation and Context

The Artemis documentation states that REST endpoints should be written in kebab case: [documentation](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#create-a-local-branch)

### Description

This PR adds a test in the `ResourceArchitecureTest.java`, to check if all REST endpoints use kebab-case (and not camel case e.g.), and fixes the endpoints which are not defined correctly.

### Steps for Testing

In theory functionality wise nothing should have changed, as only some endpoint URLs were refactored/renamed.
You can check if features related to the following refactored endpoints still work correctly:
1. LTI onlineCourseConfiguration
2. lockedSubmissions, withLatestResult
3. exerciseGroups
4. cleanupBuildPlan, buildArtifact
5. aeolus templateScripts

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
	- Updated URL paths for various REST API endpoints to improve consistency and readability, including:
		- `templateScripts` to `template-scripts`
		- `onlineCourseConfiguration` to `online-course-configuration`
		- `lockedSubmissions` to `locked-submissions`
		- `exerciseGroups` to `exercise-groups`
		- `cleanupBuildPlan` to `cleanup-build-plan`
		- `buildArtifact` to `build-artifact`

- **Bug Fixes**
	- Corrected URL references in test cases to align with the updated endpoint naming conventions.

- **Documentation**
	- Updated documentation to reflect the revised URL structures for enhanced usability.

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java:143]: The `ResourceArchitectureTest` already contains a list of these endpoint annotations that you can reuse here. I also think that this class would be a better place for this test.

Comment (SimonEntholzer) [reply src/test/java/de/tum/in/www1/artemis/architecture/ArchitectureTest.java:143]: Ah true, that's absolutely right :+1:

File: src/test/javascript/spec/service/participation.service.spec.ts
Comment (Strohgelaender) [inline src/test/javascript/spec/service/participation.service.spec.ts:93]:
```suggestion
        service.cleanupBuildPlan(participationDefault).subscribe((resp) => expect(resp).toMatchObject(participationDefault));
```

File: src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java
Comment (JohannesStoehr) [inline src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java:77]: This test currently fails:
> de.tum.in.www1.artemis.web.rest.UserResource.getVcsAccessToken(java.lang.Long) violates rule to use kebab case for rest call annotations

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java
Comment (SimonEntholzer) [reply src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java:77]: yep, the refactor for these is in https://github.com/ls1intum/Artemis/pull/9182. When it is merged it will fix this here. I removed the ready for review label until then

Comment (SimonEntholzer) [reply src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java:77]: That should be fixed now :+1:

File: src/main/java/de/tum/in/www1/artemis/web/rest/AeolusTemplateResource.java
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/AeolusTemplateResource.java:91]: Ah seems like I missed that one, as it was incorrect in the first place :+1: thanks for noticing, I fixed the comment

**Review by iyannsch — APPROVED** [review]
Code LGTM

File: src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java
Comment (N0W0RK) [inline src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java:92]: Why don't you match the complete path?

File: src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java
Comment (N0W0RK) [inline src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java:39]: I believe the outer capture group is unnecessary and the brackets could be removed.

**Review by N0W0RK — CHANGES_REQUESTED** [review]
I see a couple issues with the test. I left comments at the appropriate locations.

File: src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java
Comment (SimonEntholzer) [reply src/test/java/de/tum/in/www1/artemis/architecture/ResourceArchitectureTest.java]: no idea why I didn't think of that. I made it into one regex, which is quite huge, but the code using it is simpler.

**Review by iyannsch — DISMISSED** [review]

**Review by N0W0RK — CHANGES_REQUESTED** [review]
Looks good to me.
Is there a reason why there are some non capturing groups? I would suggest having all groups capturing for easier readability.

**Comment by SimonEntholzer** [comment]
> Looks good to me. Is there a reason why there are some non capturing groups? I would suggest having all groups capturing for easier readability.

I found a simpler regex without the non-capturing groups :+1:

File: src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java
Comment (iyannsch) [inline src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java:1]: ChatGPT considers those expressions not to be equivalent 😅 The old pattern enforces kebab-case formatting `([a-z0-9]+(?:-[a-z0-9]+)*)`, while the new pattern allows periods and does not strictly enforce kebab-case. For example `"segment.with.period/"` seems possible now, is that intended?

**Review by iyannsch — CHANGES_REQUESTED** [review]

File: src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java
Comment (N0W0RK) [inline src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java:38]: I think I have found a valid regex:

`(\\.?[a-z0-9]+(-[a-z0-9]+)*|\\{[^}]+\\})`:
Matches either a kebab case string with optional period at the beginning. (`\\.?[a-z0-9]+(-[a-z0-9]+)*`)
Or a path parameter (`\\{[^}]+\\}`)

`(/(([a-z0-9]+(-[a-z0-9]+)*|\\{[^}]+\\})))*`:
Same story as above. However everything is preceded by a `/` and the optional period is removed.
This can be repeated as much as needed.

`(\\.json|/\\*)?`:
Optionally allows the path to either end with `.json` or with `/*`

```suggestion
    private static final Pattern KEBAB_CASE_PATH_PATTERN = Pattern.compile("^(\\.?[a-z0-9]+(-[a-z0-9]+)*|\\{[^}]+\\})(/(([a-z0-9]+(-[a-z0-9]+)*|\\{[^}]+\\})))*(\\.json|/\\*)?$");
```

**Review by N0W0RK — CHANGES_REQUESTED** [review]

File: src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java
Comment (SimonEntholzer) [reply src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java]: Good suggestion, that's probably the cleanest solution :+1: I added it

File: src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java
Comment (SimonEntholzer) [reply src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java:1]: True, that wasn't as well defined as it could have been. Some paths have periods (somehting.json e.g.), but yea, allowing dots in general is probably not such a good idea.

**Review by iyannsch — DISMISSED** [review]
reapprove code

**Review by N0W0RK — DISMISSED** [review]

**Review by janthoXO — DISMISSED** [review]
looks good
does it make sense to add a test for the client side as well? or maybe something that gives at least a compiler warning when using an URL in a httpclient method that is not in kebab case

**Comment by SimonEntholzer** [comment]
> looks good does it make sense to add a test for the client side as well? or maybe something that gives at least a compiler warning when using an URL in a httpclient method that is not in kebab case

Might be useful, but I'm not sure on how that would be achieved, and  I'm also not sure if it would be worth the effort. Especially because this test here, will not allow any corresponding endpoints at the server, so the client couldn't really use such an endpoint anyway.

File: src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java
Comment (pzdr7) [inline src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java:38]: Regex makes sense.

One small question: I assume the `.json` at the end means exactly ".json" for JSON accesses. Don't we have to escape the dot, then?

Otherwise, we'd match strings like

```
foo/barXjson
```

Comment (SimonEntholzer) [reply src/test/java/de/tum/cit/aet/artemis/architecture/ResourceArchitectureTest.java:38]: good point, that's true. I'll fix that

**Review by pzdr7 — DISMISSED** [review]
Thanks for the quick change. Code

**Review by raffifasaro — DISMISSED** [review]
Code LGTM

**Review by N0W0RK — DISMISSED** [review]

**Review by asliayk — APPROVED** [review]
code lgtm

**Review by iyannsch — APPROVED** [review]
Reapprove code

**Review by Strohgelaender — APPROVED** [review]
code

**Comment by janthoXO** [comment]
> > looks good does it make sense to add a test for the client side as well? or maybe something that gives at least a compiler warning when using an URL in a httpclient method that is not in kebab case
> 
> Might be useful, but I'm not sure on how that would be achieved, and I'm also not sure if it would be worth the effort. Especially because this test here, will not allow any corresponding endpoints at the server, so the client couldn't really use such an endpoint anyway.

yeah that's true and for external apis the check does not make sense anyways.
Otherwise the code looks good, so approve from my side

**Review by janthoXO — DISMISSED** [review]

**Review by coolchock — APPROVED** [review]
tested on ts2, endpoints work fine

**Review by BBesrour — APPROVED** [review]
Tested on TS2, tested features related to these topics (not sure how to test the others)

- withLatestResult
- exerciseGroups
- buildArtifact
- aeolus templateScripts

**Review by iyannsch — APPROVED** [review]
reapprove after merge

**Review by N0W0RK — APPROVED** [review]
Reapprove after merge

**Review by pzdr7 — APPROVED** [review]
Reapprove

**Review by coolchock — APPROVED** [review]
re-approve

**Review by BBesrour — APPROVED** [review]
re-approve after merge request

**Pull request merged**
